From: mrapple <tony@oc.tc>
Date: Fri, 2 May 2014 20:48:11 -0500
Subject: [PATCH] Cache object to string translations in JsonList

With 1.7 the ban, white, and op lists all use JsonList. While this is nice, when checking if an object is in a JsonList the object is converted to a string. Since most of these objects contain UUIDs and UUID.toString() is a generally expensive call, we need to cache the string representation of the object.

Without this change, player.isOp() and player.hasPermission() are far more expensive than they were in 1.6.

diff --git a/src/main/java/net/minecraft/server/GameProfileBanList.java b/src/main/java/net/minecraft/server/GameProfileBanList.java
--- a/src/main/java/net/minecraft/server/GameProfileBanList.java
+++ b/src/main/java/net/minecraft/server/GameProfileBanList.java
@@ -33,7 +33,7 @@ public class GameProfileBanList extends JsonList<GameProfile, GameProfileBanEntr
     }
 
     protected String b(GameProfile gameprofile) {
-        return gameprofile.getId().toString();
+        return super.a(gameprofile); // CraftBukkit - use super method that uses cache
     }
 
     public GameProfile a(String s) {
diff --git a/src/main/java/net/minecraft/server/JsonList.java b/src/main/java/net/minecraft/server/JsonList.java
--- a/src/main/java/net/minecraft/server/JsonList.java
+++ b/src/main/java/net/minecraft/server/JsonList.java
@@ -29,6 +29,12 @@ import org.apache.commons.io.IOUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+// SportBukkit start
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+// SportBukkit end
+
 public class JsonList<K, V extends JsonListEntry<K>> {
 
     protected static final Logger a = LogManager.getLogger();
@@ -50,6 +56,16 @@ public class JsonList<K, V extends JsonListEntry<K>> {
         }
     };
 
+    // CraftBukkit - add cache of object -> string translations
+    LoadingCache<K, String> pastTranslations = CacheBuilder.newBuilder()
+        .maximumSize(500)
+        .build(new CacheLoader<K, String>() {
+                   public String load(K obj) {
+                       return obj.toString();
+                   }
+               }
+        );
+
     public JsonList(File file) {
         this.c = file;
         GsonBuilder gsonbuilder = (new GsonBuilder()).setPrettyPrinting();
@@ -112,7 +128,15 @@ public class JsonList<K, V extends JsonListEntry<K>> {
     }
 
     protected String a(K k0) {
-        return k0.toString();
+        // SportBukkit start - use cache
+        String result = null;
+
+        try {
+            result = this.pastTranslations.get(k0);
+        } catch (Exception e) {}
+
+        return result;
+        // SportBukkit end
     }
 
     protected boolean d(K k0) {
diff --git a/src/main/java/net/minecraft/server/OpList.java b/src/main/java/net/minecraft/server/OpList.java
--- a/src/main/java/net/minecraft/server/OpList.java
+++ b/src/main/java/net/minecraft/server/OpList.java
@@ -41,7 +41,7 @@ public class OpList extends JsonList<GameProfile, OpListEntry> {
     }
 
     protected String c(GameProfile gameprofile) {
-        return gameprofile.getId().toString();
+        return super.a(gameprofile); // SportBukkit - use super method that uses cache
     }
 
     public GameProfile a(String s) {
diff --git a/src/main/java/net/minecraft/server/WhiteList.java b/src/main/java/net/minecraft/server/WhiteList.java
--- a/src/main/java/net/minecraft/server/WhiteList.java
+++ b/src/main/java/net/minecraft/server/WhiteList.java
@@ -33,7 +33,7 @@ public class WhiteList extends JsonList<GameProfile, WhiteListEntry> {
     }
 
     protected String b(GameProfile gameprofile) {
-        return gameprofile.getId().toString();
+        return super.a(gameprofile); // CraftBukkit - use super method that uses cache
     }
 
     public GameProfile a(String s) {
